pymc4/variational/approximations.py

```diff
@@ -0,0 +1,212 @@
+    """Full Rank Automatic Differential Variational Inference(Full Rank ADVI)."""
+
+    def _build_loc(self):
+        pass
+
+    def _build_cov_matrix(self):
+        pass
+
+    def _build_posterior(self):
+        pass
+
+
+class LowRank(Approximation):
+    """Low Rank Automatic Differential Variational Inference(Low Rank ADVI)."""
+
+    def _build_loc(self):
+        pass
+
+    def _build_cov_matrix(self):
+        pass
+
+    def _build_posterior(self):
+        pass
+
+
+def fit(
+    model: Optional[Model] = None,
+    method: Union[str, MeanField] = "advi",
+    num_steps: int = 10000,
+    sample_size: int = 1,
+    random_seed: Optional[int] = None,
+    optimizer=None,
+    **kwargs,
+):
+    """
+    Fit an approximating distribution to log_prob of the model.
+
+    Parameters
+    ----------
+    model : Optional[:class:`Model`]
+        Model to fit posterior against
+    method : Union[str, :class:`Approximation`]
+        Method to fit model using VI
+
+        - 'advi' for :class:`MeanField`
+        - 'fullrank_advi' for :class:`FullRank`
+        - 'lowrank_advi' for :class:`LowRank`
+        - or directly pass in :class:`Approximation` instance
+    num_steps : int
+        Number of iterations to run the optimizer
+    sample_size : int
+        Number of Monte Carlo samples used for approximation
+    random_seed : Optional[int]
+        Seed for tensorflow random number generator
+    optimizer : TF1-style or TF2-style optimizer
+        Tensorflow optimizer to use
+    kwargs : Optional[Dict[str, Any]]
+        Pass extra non-default arguments to
+        ``tensorflow_probability.vi.fit_surrogate_posterior``
+
+    Returns
+    -------
+    ELBO : list|dict
+        Negative ELBO loss depending on the `trace_fn`
+    """
+    _select = dict(advi=MeanField,)
+
+    if isinstance(method, str):
+        # Here we assume that `model` parameter is provided by the user.
+        try:
+            inference = _select[method.lower()](model, random_seed)
+        except KeyError:
+            raise KeyError(
+                "method should be one of %s or Approximation instance" % set(_select.keys())
+            )
+
+    elif isinstance(method, Approximation):
+        # Here we assume that `model` parameter is not provided by the user
+        # as the :class:`Approximation` itself contains :class:`Model` instance.
+        inference = method
+
+    else:
+        raise TypeError(
+            "method should be one of %s or Approximation instance" % set(_select.keys())
+        )
+
+    if optimizer:
+        opt = optimizer
+    else:
+        opt = tf.optimizers.Adam(learning_rate=0.1)
+
+    losses = tfp.vi.fit_surrogate_posterior(
+        target_log_prob_fn=inference.target_log_prob,
+        surrogate_posterior=inference.approx,
+        num_steps=num_steps,
+        sample_size=sample_size,
+        seed=random_seed,
+        optimizer=opt,
+        **kwargs,
+    )
+
+    return losses
```

Comment:
I can't get approximate posterior there, only loss, in case I do not explicitly pass approximation. As a result, I would expect Approximation class that is possible to give trace samples.

Reply:
Resolved in a58a412